SI-5779: Wrong warning message (comparing Number)

som-snytt opened the pull request:
Probably this bug won't exist in some future version of Scala.  

It's hardly a bug; more like a gnat.

BoxesRuntime knows how to compare java.lang.Number, so we must not warn.

som-snytt committed:
SI-5779: Wrong warning message (comparing values of types Float/Double and Number using `==' will always yield false)

BoxesRuntime knows how to compare java.lang.Number, so we must not warn.

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/146/

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1123,10 +1126,11 @@ abstract class RefChecks extends InfoTransform with reflect.internal.transform.R
         }
-        else if (isNumeric(receiver)) {
-          if (!isNumeric(actual) && !forMSIL)
+        else if (isNumeric(receiver) || (!forMSIL && isJavaNumeric(receiver))) {
```
these conditions are getting pretty unwieldy -- could you please simplify or elucidate with a comment?
think of the poor reviewer and future-selves reading this code

adriaanm commented:
thanks for the fix! it looks like something that should go in but we have limited reviewing bandwidth
could you please help us out juuuust a bit more and make it obviously correct (either by introducing variables that name the conditions, comments (including a comment here?))

som-snytt commented:
OK.  We're sensitive to the bandwidth issue.  At a certain point, a low-priority or no-priority pull req is just inefficient or a waste of time.

I'll comment.  (It begs for refactor that is out of scope.  No doubt Paul will do that while on sabbatical.)  (I picked this bug because I just fixed a bug in its neighborhood.)

som-snytt committed:
SI-5779: Wrong warning message (comparing Number)

Massage the predicates for readability.

Thanks to Adriaan for the push.  As an outsider, one tries
to tread lightly by minimizing the number of chars changed
in the source, which is the wrong optimization.

This code remains as it was (not amenable to extension).
This commit also deletes the flag I'd introduced to say that an
exhaustive check had already been done.  For the record, I
renamed it to something less silly before I zapped it.

som-snytt commented:
The change amounts to: add isJavaNumber to isAnyNumber (as commented).

This commit isn't squashed; the diff would be cleaner.

adriaanm commented:
LGTM -- thanks! waiting for the build bot to churn through the second commit and I'll merge (unless someone beats me to it, that'd be okay too)

adriaanm commented:
build ok: https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-testsuite-linux-opt/152/
